Added documentation about the coverage of ESH Coding Guidelines

petarnv opened the pull request:
I saw this opened issue - https://github.com/openhab/static-code-analysis/issues/23. I have added an `ESH Guidelines Covered` section to the README.md file.

kaikreuzer commented on README.md:
```diff
@@ -224,7 +224,49 @@ For more information: https://pmd.github.io/pmd-5.4.1/customizing/rule-guideline
+# ESH Guidelines Covered
+
+## A. Code Style
+1. The Java naming conventions should be used. - https://github.com/openhab/static-code-analysis/blob/master/src/main/resources/rulesets/checkstyle/rules.xml#L80
```
Maybe better make this to a link for better readability: 
```
1. [The Java naming conventions should be used.](https://github.com/openhab/static-code-analysis/blob/master/src/main/resources/rulesets/checkstyle/rules.xml#L80)
```

petarnv replied:
Done.

kaikreuzer commented on README.md:
```diff
@@ -224,7 +224,49 @@ For more information: https://pmd.github.io/pmd-5.4.1/customizing/rule-guideline
+3. Every class, interface and enumeration should have JavaDoc describing its purpose and usage. - https://github.com/openhab/static-code-analysis/blob/master/src/main/resources/rulesets/checkstyle/rules.xml#L66
+4. Every class, interface and enumeration must have an @author tag in its JavaDoc for every author that wrote a substantial part of the file. - `Work in Progress`
+5. Every constant, field and method with default, protected or public visibility should have JavaDoc (optional, but encouraged for private visibility as well). - `Not covered yet`
+6. Code must be formatted using the provided code formatter and clean up settings. They are set up automatically by the official IDE setup. - `Not covered yet`
```
This should probably be split up - we have dozens of code formatter settings, and we won't (and don't want) to maintain them separately here as well in all details.

With #12, we will have the most common issue covered already - we should add a similar check for xml and json files (where we actually expect tabs, no spaces). Having these checks in place, I would consider A.6 to be checked well enough.

petarnv replied:
It is now split on three parts - formatting for java, xml and json.

maggu2810 commented on README.md:
```diff
@@ -224,7 +224,49 @@ For more information: https://pmd.github.io/pmd-5.4.1/customizing/rule-guideline
+3. Every bundle must contain a build.properties file, which lists all resources that should end up in the binary under bin.includes. - `Work in Progress`
+4. The manifest must not contain any “Require-Bundle” entries. Instead, “Import-Package” must be used. - `Opened PR` (https://github.com/openhab/static-code-analysis/pull/19)
+5. The manifest must not export any internal package. - https://github.com/openhab/static-code-analysis/blob/master/src/main/resources/rulesets/checkstyle/rules.xml#L40
+6. The manifest must not have any version constraint on package imports, unless this is thoughtfully added. Note that Eclipse automatically adds these constraints based on the version in the target platform, which might be too high in many cases. - `Work in Progress`
```
> The manifest must not have any version constraint on package imports, unless this is thoughtfully added.

How do you plan to check if a version constraint is added thoughtfully? :wink:

petarnv replied:
Well, we can add configuration properties for packages, that are usually imported with a version (like `org.junit`, `org.apache.commons`, `org.slf4j` and probably others) and these packages will not be considered a violation, even if they are imported with a version.

Also, if there is a bundle, that really needs to import a package with a version, it can be added to the suppressions.xml file (https://github.com/openhab/static-code-analysis/blob/master/src/main/resources/rulesets/checkstyle/suppressions.xml) and it will not be checked.

maggu2810 commented on README.md:
```diff
@@ -224,7 +224,49 @@ For more information: https://pmd.github.io/pmd-5.4.1/customizing/rule-guideline
+1. Every bundle must contain a Maven pom.xml with a version and artifact name that is in sync with the manifest entry. The pom.xml must reference the correct parent pom (which is usually in the parent folder). - `Work in Progress`
+2. Every bundle must contain an about.html file, providing license information. - `Work in Progress`
+3. Every bundle must contain a build.properties file, which lists all resources that should end up in the binary under bin.includes. - `Work in Progress`
+4. The manifest must not contain any “Require-Bundle” entries. Instead, “Import-Package” must be used. - `Opened PR` (https://github.com/openhab/static-code-analysis/pull/19)
```
:+1: 
But I assume this could be a `hard work` for the model bundles.

Should we exclude test bundles?

petarnv replied:
Yes, we can exclude the model bundles. But I'm not sure what do you mean when you say that checking the model bundles will be hard work?

kaikreuzer replied:
@maggu2810 For the model bundles, we should be able to deactivate this check in the pom.xml (@petarnv, I understood that this is possible for individual checks, right?)

petarnv commented:
I updated the PR. Now for every implemented check, its severity/priority can be seen. 

I have also added a section for all checks that we have from Checkstyle, PMD and Findbugs with links to their configuration files and their documentation.
